Pre restart probe

bashtanov opened the pull request:
This is to implement a broker-level API endpoint to be called before restarting the broker node. It returns lists of  partitions affected by one of the risks in case the node goes down. Partition risks assessed are:
- going leaderless
- refusing to produce with full acks
- losing data produced with acks=1
- partition with replication factor 1 going completely unavailable (separate, as in a sense expected).

See [RFC](https://docs.google.com/document/d/1r10L9ruSb2RavdnqOfujtVuWFmNCwLYTkuB5_T5dsFA/edit?tab=t.0#heading=h.3lbgkvc963x) for details.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Features
* pre-restart check API endpoint
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
cluster/tests/health_monitor: bazelize

bashtanov committed:
cluster/health_monitor: types for restart probes

bashtanov committed:
c/health_monitor/backend: calculate restart risks and post-restart score

bashtanov committed:
c/health_monitor/backend: remove invalid comment

When collecting and aggregating cluster health data we don't treat raft0
leader specially. Each node collects data from itself and all other
nodes directly.

bashtanov committed:
c/health_monitor/backend: check if all nodes collect restart risks

And store it in a boolean flag. That's to distinguish between no known
risks and poor risk awareness on old nodes. We'll need it when asked for
risks by admin API.

bashtanov committed:
features/table: add feature for node restart risk assessment

bashtanov committed:
c/health_monitor/frontend: route cur. node pre/post-restart probes to BE

bashtanov committed:
rp/admin/api-doc/broker: declare pre-/post-restart probe endpoints

bashtanov committed:
rp/admin/server: implement /v1/broker/{pre,post}_restart_probe endpoints

bashtanov committed:
tests/services/admin: query broker/{pre,post}_restart_probe

bashtanov commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--6a4a9d07-08a4-4887-9b32-f1f42a96e193-->
<!--TEST RESULTS-->
<!--3ff294e6-5c68-423a-a500-1c0299a12d8c-->
<!--TEST RESULTS-->
<!--eff27368-5fc3-4c1c-8e3c-fda6538f7ec2-->
<!--TEST RESULTS-->
<!--b51b04da-096d-419e-b85e-d7553e5bab34-->
<!--TEST RESULTS-->
<!--5db52634-0ba5-43ae-8ee1-3ed44f25bbe4-->
<!--TEST RESULTS-->
<!--fd7761df-c432-47db-ae4e-f622cad433c4-->
<!--TEST RESULTS-->
<!--3e5f3043-3fc8-4c35-b666-93ab9da11759-->
<!--TEST RESULTS-->
<!--a8be47e6-7362-4439-9df7-ababdf0a51a2-->
<!--TEST RESULTS-->
<!--19c4675c-7bc3-46c4-adcd-93e495bcdfce-->
<!--TEST RESULTS-->
<!--b50340ca-f687-473f-bf8f-fffe65bc5a6d-->
<!--TEST RESULTS-->
<!--462138e0-951f-4327-a4a2-0faf04c3a156-->
<!--TEST RESULTS-->
<!--756409d2-d2ff-4e89-a7ad-0ee1dc54de86-->
<!--TEST RESULTS-->
<!--aa65cabd-6b4e-4cd1-a3e8-c9769239f10a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61165">build#61165</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.controller_log_limiting_test.ControllerLogLimitPartitionBalancerTests.test_partition_balancer_with_limits | ducktape | https://buildkite.com/redpanda/redpanda/builds/61165#019498cb-96b0-47f6-aec1-a18ac3bf99fe | FLAKY | 1/3 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_crashed_node | ducktape | https://buildkite.com/redpanda/redpanda/builds/61165#019498cb-96ae-4a54-93ef-f6b100a00c62 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=False.tick_interval=3600000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61165#019498cb-96af-4979-bebc-a71731f73b53 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recommissioning_node_finishes | ducktape | https://buildkite.com/redpanda/redpanda/builds/61165#019498cb-96b0-47f6-aec1-a18ac3bf99fe | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recommissioning_one_of_decommissioned_nodes | ducktape | https://buildkite.com/redpanda/redpanda/builds/61165#019498cb-96ae-4a54-93ef-f6b100a00c62 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61165#019498cb-96af-4979-bebc-a71731f73b53 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61165#019498cb-96af-4d70-a4bd-8e6c253b96c3 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_multiple_nodes_to_the_cluster.partition_count=20 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61165#019498cb-96af-4d70-a4bd-8e6c253b96c3 | FLAKY | 1/3 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_node_with_unavailable_node | ducktape | https://buildkite.com/redpanda/redpanda/builds/61165#019498cb-96b0-47f6-aec1-a18ac3bf99fe | FLAKY | 1/4 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_topic_hot_spots | ducktape | https://buildkite.com/redpanda/redpanda/builds/61165#019498cb-96ae-4a54-93ef-f6b100a00c62 | FLAKY | 1/4 |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group | ducktape | https://buildkite.com/redpanda/redpanda/builds/61165#019498cb-96ae-4a54-93ef-f6b100a00c62 | FLAKY | 1/2 |
| test_archiver_manager_smp_rpfixture.test_archiver_manager_smp_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61165#01949870-a347-4a79-97cd-684af0264ca1 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61311">build#61311</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb08-4915-b748-9630bc1ee562 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=False.tick_interval=3600000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb08-4915-b748-9630bc1ee562 | FLAKY | 1/3 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=False.tick_interval=5000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb09-4d6a-8137-f36f91282e3d | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_node_is_not_allowed_to_join_after_restart.new_bootstrap=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb07-470a-939e-52e7b1380963 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb08-4915-b748-9630bc1ee562 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb09-4d6a-8137-f36f91282e3d | FLAKY | 1/3 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsACLsTest.test_reassignments_with_acls | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb08-4d76-a89d-017dc3642536 | FLAKY | 1/3 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_multiple_nodes_to_the_cluster.partition_count=20 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb09-4d6a-8137-f36f91282e3d | FLAKY | 1/15 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_node_with_unavailable_node | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb07-470a-939e-52e7b1380963 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_on_demand_rebalancing.partition_count=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb09-4d6a-8137-f36f91282e3d | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_topic_hot_spots | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb08-4d76-a89d-017dc3642536 | FLAKY | 1/3 |
| rptest.tests.shard_placement_test.ShardPlacementTest.test_node_join.disable_license=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61311#0194adfe-cb07-470a-939e-52e7b1380963 | FLAKY | 1/2 |
| test_archiver_manager_smp_rpfixture.test_archiver_manager_smp_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61311#0194adba-1ddb-48cc-8916-8bc7be7f68ea | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61344">build#61344</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| cluster_decommissioning_tests_rpfixture.cluster_decommissioning_tests_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61344#0194b1b5-0815-41a3-93ed-e0265e59f1c4 | FLAKY | 1/2 |
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61344#0194b1b5-0815-41a3-93ed-e0265e59f1c4 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0ba9-448b-8af9-0622354c8d04 | FLAKY | 1/5 |
| rptest.tests.controller_log_limiting_test.ControllerLogLimitPartitionBalancerTests.test_partition_balancer_with_limits | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0ba8-4b94-95cd-5739925b2464 | FLAKY | 1/2 |
| rptest.tests.e2e_iam_role_test.AWSRoleFetchTests.test_write | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0ba8-4b94-95cd-5739925b2464 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=False.tick_interval=5000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0baa-4416-bc5a-972f6c8bf8a0 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=True.tick_interval=3600000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0ba8-4b94-95cd-5739925b2464 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_flipping_decommission_recommission.node_is_alive=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0baa-4416-bc5a-972f6c8bf8a0 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0ba9-448b-8af9-0622354c8d04 | FLAKY | 1/19 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0baa-4416-bc5a-972f6c8bf8a0 | FLAKY | 1/4 |
| rptest.tests.partition_movement_test.PartitionMovementTest.test_availability_when_one_node_down | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0ba9-4f13-acc0-43dfba5d7e31 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_multiple_nodes_to_the_cluster.partition_count=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0ba9-448b-8af9-0622354c8d04 | FLAKY | 1/3 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_multiple_nodes_to_the_cluster.partition_count=20 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0baa-4416-bc5a-972f6c8bf8a0 | FLAKY | 1/4 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_node_with_unavailable_node | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0ba8-4b94-95cd-5739925b2464 | FLAKY | 1/4 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_on_demand_rebalancing.partition_count=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0baa-4416-bc5a-972f6c8bf8a0 | FLAKY | 1/3 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_topic_hot_spots | ducktape | https://buildkite.com/redpanda/redpanda/builds/61344#0194b211-0ba9-4f13-acc0-43dfba5d7e31 | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61395">build#61395</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.datalake_dlq_test.DatalakeDLQTest.test_dlq_table_for_mixed_records.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.filesystem_catalog_mode=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61395#0194b701-a8ec-4560-8239-43c75e9747c4 | FLAKY | 1/2 |
| rptest.tests.e2e_iam_role_test.ShortLivedCredentialsTests.test_short_lived_credentials | ducktape | https://buildkite.com/redpanda/redpanda/builds/61395#0194b701-a8ec-407e-a3ac-ef1c56572b57 | FLAKY | 1/3 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTest.test_write.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61395#0194b701-a8ea-4afb-93df-9633177fc4c8 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=False.tick_interval=5000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61395#0194b701-a8ec-4560-8239-43c75e9747c4 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=True.tick_interval=5000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61395#0194b701-a8eb-4109-821d-801651a5cc9f | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_node_is_not_allowed_to_join_after_restart.new_bootstrap=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61395#0194b701-a8eb-4109-821d-801651a5cc9f | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61395#0194b701-a8ec-407e-a3ac-ef1c56572b57 | FLAKY | 1/14 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61395#0194b701-a8ec-4560-8239-43c75e9747c4 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_multiple_nodes_to_the_cluster.partition_count=20 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61395#0194b701-a8ec-4560-8239-43c75e9747c4 | FLAKY | 1/5 |
| test_archiver_manager_smp_rpfixture.test_archiver_manager_smp_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61395#0194b6bd-5f3c-44d6-b4fe-b18b088a19d4 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61402">build#61402</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| cluster_decommissioning_tests_rpfixture.cluster_decommissioning_tests_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61402#0194b862-b4ec-43c6-8d73-9650a5de4a05 | FLAKY | 1/2 |
| rptest.tests.cluster_metadata_upload_test.ClusterMetadataUploadTest.test_uploads_after_restart.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-0249-4973-9058-fa6ea6141711 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-024a-4936-b028-c46826406e59 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-024b-410b-8252-46bab85db46c | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=False.tick_interval=3600000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-0249-4973-9058-fa6ea6141711 | FLAKY | 1/3 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=False.tick_interval=5000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-024a-415f-a0d6-4fd93a567f8f | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-0249-4973-9058-fa6ea6141711 | FLAKY | 1/4 |
| rptest.tests.rack_aware_replica_placement_test.RackAwarePlacementTest.test_rack_awareness_after_node_operations.rack_layout=xxYYzz | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-024a-4936-b028-c46826406e59 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_multiple_nodes_to_the_cluster.partition_count=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-0249-4973-9058-fa6ea6141711 | FLAKY | 1/3 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_multiple_nodes_to_the_cluster.partition_count=20 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-024a-415f-a0d6-4fd93a567f8f | FLAKY | 1/15 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_node_with_unavailable_node | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-024a-4936-b028-c46826406e59 | FLAKY | 1/3 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_on_demand_rebalancing.partition_count=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-024a-415f-a0d6-4fd93a567f8f | FLAKY | 1/4 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_topic_hot_spots | ducktape | https://buildkite.com/redpanda/redpanda/builds/61402#0194b8be-024b-410b-8252-46bab85db46c | FLAKY | 1/2 |
| test_archiver_manager_smp_rpfixture.test_archiver_manager_smp_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61402#0194b862-b4eb-4721-a296-686b1783f373 | FLAKY | 1/2 |
| test_archiver_manager_smp_rpfixture.test_archiver_manager_smp_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61402#0194b862-b4ec-43c6-8d73-9650a5de4a05 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61415">build#61415</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| cluster_decommissioning_tests_rpfixture.cluster_decommissioning_tests_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61415#0194b980-e4d3-4b69-8986-66821a1a7b86 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61415#0194b9c5-03c7-4236-85da-03526d38e550 | FLAKY | 1/5 |
| rptest.tests.controller_log_limiting_test.ControllerLogLimitPartitionBalancerTests.test_partition_balancer_with_limits | ducktape | https://buildkite.com/redpanda/redpanda/builds/61415#0194b9c5-03c6-44a2-99a1-92fefc5accd6 | FLAKY | 1/4 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=False.tick_interval=3600000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61415#0194b9c5-03c6-44a2-99a1-92fefc5accd6 | FLAKY | 1/3 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_working_node.delete_topic=True.tick_interval=3600000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61415#0194b9c5-03c7-4236-85da-03526d38e550 | FLAKY | 1/3 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61415#0194b9c5-03c6-44a2-99a1-92fefc5accd6 | FLAKY | 1/19 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recycle_all_nodes.auto_assign_node_id=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61415#0194b9c5-03c6-4094-9803-6d339d4f68c3 | FLAKY | 1/3 |
| rptest.tests.rack_aware_replica_placement_test.RackAwarePlacementTest.test_rack_awareness_after_node_operations.rack_layout=xxYYzz | ducktape | https://buildkite.com/redpanda/redpanda/builds/61415#0194b9c5-03c7-4236-85da-03526d38e550 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_multiple_nodes_to_the_cluster.partition_count=20 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61415#0194b9c5-03c6-4094-9803-6d339d4f68c3 | FLAKY | 1/5 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_on_demand_rebalancing.partition_count=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61415#0194b9c5-03c6-4094-9803-6d339d4f68c3 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_topic_hot_spots | ducktape | https://buildkite.com/redpanda/redpanda/builds/61415#0194b9c5-03c8-4425-aded-184fa4385b8f | FLAKY | 1/2 |
| test_archiver_manager_smp_rpfixture.test_archiver_manager_smp_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61415#0194b980-e4d2-4df1-b39b-a52b34c01a36 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61640">build#61640</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| cluster_health_monitor_test_rpfixture.cluster_health_monitor_test_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61640#0194d8b9-857b-4090-93a5-96fc9b1685bb | FAIL | 0/2 |
| cluster_health_monitor_test_rpfixture.cluster_health_monitor_test_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61640#0194d8b9-857b-4a26-90b2-c938cb1a6072 | FAIL | 0/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61640#0194d902-85c9-49c8-b7f1-c312e47bf156 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61640#0194d914-c63e-4335-a45c-beb17af7a5ef | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61640#0194d902-85c1-4e7e-a74a-c809fa81c2b8 | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/61640#0194d902-85c1-4e7e-a74a-c809fa81c2b8 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61663">build#61663</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| cluster_health_monitor_test_rpfixture.cluster_health_monitor_test_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61663#0194dad8-4eb2-4332-957d-e6912a74a742 | FAIL | 0/2 |
| cluster_health_monitor_test_rpfixture.cluster_health_monitor_test_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61663#0194dad8-4eb2-49ea-a924-decd27d9ec0c | FAIL | 0/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61663#0194db1c-c64d-4f9d-8ebc-16f320c631d3 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61663#0194db38-67e1-43c6-b229-302e05389f98 | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/61663#0194db38-67e1-43c6-b229-302e05389f98 | FLAKY | 1/5 |
| rptest.tests.maintenance_test.MaintenanceTest.test_maintenance_sticky.use_rpk=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61663#0194db1c-c64d-4f9d-8ebc-16f320c631d3 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61663#0194db38-67e2-4ce6-830a-18a650fcb182 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/61663#0194db38-67e2-4ce6-830a-18a650fcb182 | FLAKY | 1/3 |
| rptest.transactions.stream_verifier_test.StreamVerifierTest.test_simple_produce_consume_txn_with_add_node | ducktape | https://buildkite.com/redpanda/redpanda/builds/61663#0194db38-67e1-43c6-b229-302e05389f98 | FLAKY | 1/2 |
| storage_single_thread_rpunit.storage_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61663#0194dad8-4eb2-4332-957d-e6912a74a742 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61676">build#61676</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61676#0194dc2b-f880-4fb9-b2e1-29772d10f660 | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/61676#0194dc2b-f87e-4e7b-9ccc-12f1d2a95ee4 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61694">build#61694</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61694#0194dd76-2ba0-4abc-859b-b6145f177006 | FLAKY | 1/2 |
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_with_transactions.cloud_storage_enabled=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61694#0194dd50-a336-4e7e-802d-32722cbcb748 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62230">build#62230</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.crash_loop_checks_test.CrashLoopChecksTest.test_crash_report_with_signal.signo=Signals.SIGSEGV.signal_shard=0 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62230#01953a8a-f93d-4bf9-8108-be9f1a5f684f | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_mount_inexistent | ducktape | https://buildkite.com/redpanda/redpanda/builds/62230#01953a8a-f93f-4ca9-a1c6-bcd3bc3cde39 | FLAKY | 1/2 |
| test_kafka_protocol_unit_rpunit.test_kafka_protocol_unit_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62230#01953a46-e807-45c0-a7b7-20b691c95eaf | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62488">build#62488</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.availability_test.AvailabilityTests.test_recovery_after_catastrophic_failure | ducktape | https://buildkite.com/redpanda/redpanda/builds/62488#01954f92-ef2f-4c65-80d8-e56a87241e21 | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionTest.test_compaction.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62488#01954f7f-4290-47e1-afad-ba7d047069ca | FLAKY | 1/3 |
| rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact | ducktape | https://buildkite.com/redpanda/redpanda/builds/62488#01954f92-ef2f-4c65-80d8-e56a87241e21 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62488#01954f7f-4292-4129-8b82-a79509a01841 | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62534">build#62534</a></summary>


| test_id                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   |
|:---------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| distributed_kv_stm_tests_rpunit.distributed_kv_stm_tests_rpunit                                                            | unit        | https://buildkite.com/redpanda/redpanda/builds/62534#01955b80-29d0-4075-b8fa-92b49aa6d56e | FLAKY         | 1/2      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62534#01955bdb-e11a-4c49-bed2-4ec267f97905 | FLAKY         | 1/2      |
| rptest.tests.crash_loop_checks_test.CrashLoopChecksTest.test_crash_report_with_signal.signo=Signals.SIGABRT.signal_shard=0 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62534#01955bdb-e11b-47e4-8422-bb7f09d06ef3 | FLAKY         | 1/2      |
</details>
<!--/body-->

bashtanov committed:
raft/types/follower_metrics: implement output operator

bashtanov committed:
r/consensus: add logging for consensus::get_follower_metrics

bashtanov committed:
rp/admin/partition: fix error messages

bashtanov committed:
tests/maintenance: leadership disbalance is <3x after rolling restart

A rolling restart, or just a rolling leader eviction by
maintentance mode, will cause leader disbalance, assuming leader
balancer is no there to fix it.
Let P be the number of partitions and N number of nodes.
For reasonably large P and even original leader distribution
(~P/N leaders on each node), this process should put as many as
e x P/N leaders onto the node restarted last (other nodes receive less
leaders). Add a test to show this in practice.

bashtanov commented:
/dt

bashtanov committed:
tests/maintenance: separate utility mixin

separate maintenance utilities into a mixin so that we can trigger
maintenance mode in other tests where needed

bashtanov commented:
/dt

ztlpn commented on src/v/cluster/health_monitor_types.h:
```diff
@@ -339,6 +419,21 @@ struct cluster_health_report
+        void push(const model::topic_namespace&, model::partition_id);
+
+    private:
+        static constexpr size_t MAX_SIZE = 128;
```
Would be great to have this configurable by a query param. This way automatic checks that only care about absence of risk can request with a small limit, and for debug purposes we'll be able to request with a bigger limit.

ztlpn commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -490,6 +490,86 @@
+            "operations": [
+                {
+                    "method": "POST",
+                    "summary": "Block a raft operation on this node",
```
Nit: maybe add a note that this is intended for testing/fault injection purposes

ztlpn commented on src/v/redpanda/admin/api-doc/broker.json:
```diff
@@ -470,6 +485,54 @@
+                    },
+                    "description": "Partitions with replication factor 1 with replica on current node"
+                },
+                "full_acks_produce_unavail" : {
```
nit: let's not shorten the word, it doesn't help much

bashtanov replied:
a 29-char identifier with 80 characters per line limit? okay

ztlpn commented on src/v/raft/consensus.h:
```diff
@@ -946,6 +951,8 @@ class consensus {
 
     replication_monitor _replication_monitor;
 
+    bool append_entries_blocked = false;
```
class members should be prefixed with `_`

ztlpn replied:
nit: a comment mentioning that this is for fault injection would be nice

ztlpn commented on src/v/redpanda/admin/partition.cc:
```diff
@@ -532,6 +532,37 @@ admin_server::force_set_partition_replicas_handler(
+    // Strange situation, but need to check it
+    if (!shard) {
+        throw ss::httpd::server_error_exception(fmt_with_ctx(
+          fmt::format, "Can not find shard for partition {}", ntp.tp));
```
nit: why not just ntp as in other handlers?

bashtanov replied:
will fix

ztlpn commented on src/v/redpanda/admin/server.h:
```diff
@@ -537,6 +539,8 @@ class admin_server {
     ss::future<ss::json::json_return_type>
       force_set_partition_replicas_handler(std::unique_ptr<ss::http::request>);
+    ss::future<ss::json::json_return_type>
+    toggle_block_raft_op(std::unique_ptr<ss::http::request>, bool block);
```
nit: bool_class would be nice

bashtanov replied:
Do you mean one local to server.h/cc only? Using it all the way through will come with more header dependencies.

ztlpn replied:
I mean anywhere we call it with explicit true/false. Though indeed may be not worth the hassle if there is no good place to put it. I wonder if we should just have a generic `enabled` bool class somewhere in `model/` :)

ztlpn commented on src/v/cluster/health_monitor_types.h:
```diff
@@ -339,6 +419,21 @@ struct cluster_health_report
 
+struct node_restart_risk_report {
+    struct affected_partitions {
+        chunked_vector<model::ntp> v;
```
nit: a longer name would be nice

ztlpn commented on src/v/redpanda/admin/server.cc:
```diff
@@ -2645,6 +2645,44 @@ admin_server::reset_crash_tracking(std::unique_ptr<ss::http::request>) {
+        co_await throw_on_error(*req, maybe_res.error(), model::controller_ntp);
+        vassert(false, "the line above should have thrown");
+    }
+    auto& res = maybe_res.value();
```
nit: const?

ztlpn commented on src/v/cluster/health_monitor_types.h:
```diff
@@ -86,9 +140,32 @@ struct node_state
+    friend bool operator==(const node_restart_risk&, const node_restart_risk&)
+      = default;
+};
+struct leader_status
```
I like the idea of packing the leader-related part of partition status into a single struct, but I don't like that it is specific to restart risks. Can we store more general info there (like the number of live replicas + lists of underreplicated/offline replicas)? This seems more generally useful and followers can do risk calculations themselves.

Also maybe we can call this struct "problematic followers" or something and skip filling it in the common case of fully replicated partition, to save on storage costs.

bashtanov replied:
> Can we store more general info there (like the number of live replicas + lists of underreplicated/offline replicas)

Sounds reasonable. I'll think about it.

> skip filling it in the common case of fully replicated partition

In case of a fully replicated partition the vector will be empty, the only way to further reduce the size would be to store a pointer instead.

bashtanov replied:
I only think that at least in case of nodes lagging or down we should send out the _list_ of in-sync followers as well. That's because on the node queried we need to be able to reliably figure out local replica status, ideally in a way consistent with the leader's info. WDYT?

ztlpn replied:
> In case of a fully replicated partition the vector will be empty, the only way to further reduce the size would be to store a pointer instead.

I mean put a nullopt in the top-level field itself.

> I only think that at least in case of nodes lagging or down we should send out the list of in-sync followers as well.

For simplicity we can do it I guess, but if the local replica is up and not in the lagging/down list, doesn't it imply that it is in-sync? (at least within the tolerance of our other assumptions)

bashtanov replied:
> to save on storage costs

> I mean put a nullopt in the top-level field itself.

What storage costs are we concerned about? It it is RAM then wrapping a vector into optional won't reduce empty vector overhead, rather the opposite. If it is serialized data I thought we don't store it for long.

> if the local replica is up and not in the lagging/down list, doesn't it imply that it is in-sync?

I was rather worried about a situation where the queried node is so screwed that it doesn't recognize it should hold a replica for the partition. In this case there will be no risk in restarting it though, at least not w.r.t. the partition in question. So I only store the number of in-sync partitions as you suggested.

ztlpn commented on src/v/cluster/health_monitor_types.h:
```diff
@@ -86,9 +140,32 @@ struct node_state
+};
+struct leader_status
+  : serde::
+      envelope<leader_status, serde::version<3>, serde::compat_version<0>> {
```
should be version 0?

ztlpn commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +275,155 @@ void health_monitor_backend::abort_current_refresh() {
+          "current node is not part of the cluster => no risks");
+        co_return report;
+    };
+    for (const auto& [nt, partitions] : local_report_it->second->topics) {
```
we should copy report ptrs and iterate with async algorithm tools

ztlpn replied:
@bashtanov this is still not fixed, right?

mmaslankaprv replied:
tip look at the counter version

bashtanov replied:
oops, must have been lost when rebasing, will fix

ztlpn commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +275,155 @@ void health_monitor_backend::abort_current_refresh() {
+                record_risk_in_report(
+                  report, _self, *li, nt, partition_status.id);
+            } else {
+                unclaimed_partitions[nt].insert(partition_status.id);
```
is it just the follower replicas set?

bashtanov replied:
It is a set of partitions we know have a local replica but haven't seen leader entry in a remote report yet.

ztlpn commented on tests/rptest/tests/node_restart_probe_test.py:
```diff
@@ -0,0 +1,282 @@
+            self.redpanda.logger.debug(f"producing to {topic.name}")
+            num_messages = int(topic.partition_count * self.PRODUCE_BYTES /
+                               self.MSG_SIZE)
+            self.kafka_tools.produce(topic.name, num_messages, self.MSG_SIZE,
```
does this run nicely in debug? 300mb seems a bit too high

bashtanov replied:
Lower numbers make the test flaky for me. Yes I run it locally in debug.

bashtanov commented:
/dt

bashtanov commented:
/dt

asimms41 commented on src/v/redpanda/admin/api-doc/broker.json:
```diff
@@ -249,6 +249,44 @@
+            "operations": [
+                {
+                    "method": "GET",
+                    "summary": "Check whether it is safe to restart this broker",
```
```suggestion
                    "summary": "Check if it this broker has recovered after restart",
```

bashtanov replied:
ouch, this one is actually incorrect, I'll change it

asimms41 commented on src/v/redpanda/admin/api-doc/broker.json:
```diff
@@ -249,6 +249,44 @@
+                        "name": "limit",
+                        "schema": {"type": "integer"},
+                        "required": false,
+                        "description": "Limit the number of partitions listed for each risk type; defaults to 128"
```
```suggestion
                        "description": "Limit the number of partitions listed for each risk type (default is 128)"
```

asimms41 commented on src/v/redpanda/admin/api-doc/broker.json:
```diff
@@ -249,6 +249,44 @@
+            "operations": [
+                {
+                    "method": "GET",
+                    "summary": "Check whether it is safe to restart this broker",
```
```suggestion
                    "summary": "Check if it is safe to restart this broker",
```

asimms41 commented on src/v/redpanda/admin/api-doc/broker.json:
```diff
@@ -470,6 +508,64 @@
+            "properties" : {
+                "risks": {
+                    "type": "restart_risks",
+                    "description": "Partitions affected by current node restart, arranged by risk type"
```
```suggestion
                    "description": "Partitions affected by the current node restart, arranged by risk type"
```

asimms41 commented on src/v/redpanda/admin/api-doc/broker.json:
```diff
@@ -470,6 +508,64 @@
+            "properties" : {
+                "load_reclaimed_pc": {
+                    "type": "int",
+                    "description": "A measure of how much load the broker has reclaimed after restart. Currently implemented as percentage of in-sync replicas. 0 to 100."
```
```suggestion
                    "description": "Measure the load the broker has reclaimed after a restart as a percentage of in-sync replicas. 0 to 100."
```

asimms41 commented on src/v/redpanda/admin/api-doc/broker.json:
```diff
@@ -470,6 +508,64 @@
+        },
+        "restart_risks": {
+            "id": "restart_risks",
+            "description": "Partitions affected by current node restart, arranged by risk type (each list truncated at 128 entries)",
```
```suggestion
            "description": "Partitions affected by the current node restart, arranged by risk type (each list is truncated according to the limit specified in the request)",
```

bashtanov replied:
And this one as well

asimms41 commented on src/v/redpanda/admin/api-doc/broker.json:
```diff
@@ -470,6 +508,64 @@
+                        "type": "string",
+                        "description": "ntp"
+                    },
+                    "description": "Partitions with replication factor 1 with replica on current node"
```
```suggestion
                    "description": "Partitions with replication factor 1 that have a replica on current node"
```

asimms41 commented on src/v/redpanda/admin/api-doc/broker.json:
```diff
@@ -470,6 +508,64 @@
+                        "type": "string",
+                        "description": "ntp"
+                    },
+                    "description": "Partitions that may reject produce requests with acks=-1 if current node is restarted"
```
```suggestion
                    "description": "Partitions that may reject produce requests (with acks=-1) if the current node is restarted"
```

asimms41 commented on src/v/redpanda/admin/api-doc/broker.json:
```diff
@@ -470,6 +508,64 @@
+                        "type": "string",
+                        "description": "ntp"
+                    },
+                    "description": "Partitions that may reject consume and produce requests if current node is restarted"
```
```suggestion
                    "description": "Partitions that may reject consume and produce requests if the current node is restarted"
```

asimms41 commented on src/v/redpanda/admin/api-doc/broker.json:
```diff
@@ -470,6 +508,64 @@
+                        "type": "string",
+                        "description": "ntp"
+                    },
+                    "description": "Partitions that may lose data produced with acks=1 if current node is restarted"
```
```suggestion
                    "description": "Partitions that may lose data produced (with acks=1) if the current node is restarted"
```

bashtanov replied:
Typically data is produced with acks=-1, producing with acks=1 is seen as a less safe compromise for better performance. Should we nevertheless keep it in brackets?

asimms41 commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -490,6 +490,86 @@
+            "operations": [
+                {
+                    "method": "POST",
+                    "summary": "block a raft operation on this node; used for failure injection",
```
```suggestion
                    "summary": "Block a raft operation on this node (used for failure injection)",
```

asimms41 commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -490,6 +490,86 @@
+            "operations": [
+                {
+                    "method": "POST",
+                    "summary": "Unblock a raft operation on this node; to undo a failure injection",
```
```suggestion
                    "summary": "Unblock a raft operation on this node (undo a failure injection)",
```

asimms41 reviewed (dismissed):
A few suggestions

ztlpn commented on tests/rptest/services/admin.py:
```diff
@@ -980,6 +980,23 @@ def get_broker(self, id, node=None):
         """
         return self._request('get', f"brokers/{id}", node=node).json()
 
+    def get_broker_pre_restart_probe(self, limit=None, node=None):
```
I think we should force the node param here, it doesn't make sense to ask a random node

ztlpn commented on src/v/cluster/health_monitor_types.h:
```diff
@@ -86,9 +138,26 @@ struct node_state
+  : serde::
+      envelope<followers_stats, serde::version<0>, serde::compat_version<0>> {
+    // all distinct
+    int16_t in_sync;
```
nit: default initializer

ztlpn commented on src/v/cluster/health_monitor_types.h:
```diff
@@ -43,6 +43,58 @@ using alive = ss::bool_class<struct node_alive_tag>;
 // An application version is a software release, like v1.2.3_gfa0d09f8a
 using application_version = named_type<ss::sstring, struct version_number_tag>;
 
+struct partition_risk
```
Looks like this struct doesn't need serialization and can even be internal to the backend

ztlpn commented on src/v/cluster/health_monitor_backend.h:
```diff
@@ -199,4 +227,21 @@ class health_monitor_backend {
     friend struct health_report_accessor;
 };
+
+struct ntp_report {
```
could be brought back into .cc

ztlpn commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +277,271 @@ void health_monitor_backend::abort_current_refresh() {
+          if (rf == replication_factor{1}) {
+              report.push(&restart_risk_report::rf1_offline, nt, pid);
+          } else {
+              report.push(&restart_risk_report::acks1_data_loss, nt, pid);
```
This is just a leaderless partition, why are we recording this risk here? I guess we cannot know for sure, all 3 replicas may have been in sync just before it lost leadership, then restarting a node is safe.

bashtanov replied:
Yes I'm reporting the worst risks as we cannot know for sure.

ztlpn commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +277,271 @@ void health_monitor_backend::abort_current_refresh() {
+    // 0 or 1 each
+    uint16_t self_out_of_sync = std::ranges::count(fs.out_of_sync, self);
+    uint16_t self_down = std::ranges::count(fs.down, self);
+    uint16_t self_in_sync = 1 - self_out_of_sync - self_down;
```
Would be nice to protect against wrapping here. Maybe assert is too strict, but an error log message if both self_out_of_sync and self_down are true seems good.

mmaslankaprv replied:
It looks like this comment is not addressed

bashtanov replied:
If they are both true (but each is just 1) the line below will fail because 65535 is not 0 nor 1. For it to give a false positive fs.down and fs.out_of_sync will need to contain the current node 65536 or 65537 times, which is fairly unlikely even if we have a bug that allows them to have it more than once.

mmaslankaprv commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +277,271 @@ void health_monitor_backend::abort_current_refresh() {
+    partition_risk restart_risk{partition_risk::c::no_risk};
+    vlog(
+      clusterlog.trace,
+      "record_risk_in_report nt={}, pid={}, self_out_of_sync={}, self_down={}, "
+      "self_in_sync={}, n_in_sync={}, n_out_of_sync={}, n_down={}",
```
would it be possible to make this log message more readable without having to refer to the code ?

bashtanov replied:
It's a trace log message I only left because I used it for debugging. I'm not sure if it is going to be useful, so I'll remove it.

mmaslankaprv commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +277,271 @@ void health_monitor_backend::abort_current_refresh() {
+
+    vlog(
+      clusterlog.trace,
+      "record_risk_in_report nt={}, pid={}, risks={}",
```
pid is usually referring to `producer_id` maybe we can render the partition and topic as `{ns}/{topic}/{partition}`

mmaslankaprv replied:
this will make it easier to grep for the logs related with partition

bashtanov committed:
ssx/async_algorithm: async_foreach with counter for a range

mmaslankaprv commented on src/v/cluster/health_monitor_types.cc:
```diff
@@ -144,19 +145,43 @@ std::ostream& operator<<(std::ostream& o, const cluster_health_report& r) {
       "{{id: {}, term: {}, leader_id: {}, revision_id: {}, size_bytes: {}, "
-      "reclaimable_size_bytes: {}, under_replicated: {}, shard: {}}}",
+      "reclaimable_size_bytes: {}, under_replicated: {}, shard: {}, "
+      "leader_info: {}}}",
```
nit: maybe this could be named `follower_stats` not to introduce a new name

mmaslankaprv commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +278,345 @@ void health_monitor_backend::abort_current_refresh() {
 
+namespace {
+struct partition_risk
+  : public serde::
+      envelope<partition_risk, serde::version<0>, serde::compat_version<0>> {
```
Are we serializing this struct ? Why do we need an envelope ?

bashtanov replied:
not anymore, I'll remove it

mmaslankaprv commented on src/v/raft/consensus.h:
```diff
@@ -543,6 +543,11 @@ class consensus {
         return _configuration_manager.has_configuration_override();
     }
 
+    void toggle_block_append_entries(bool block) {
```
This method name doesn't susggest it is a fault injection mechanism. Can we add a comment on top of that method. I think it should log a `WARN` level message

mmaslankaprv commented on src/v/raft/consensus.cc:
```diff
@@ -2129,6 +2129,15 @@ consensus::do_append_entries(append_entries_request&& r) {
         reply.may_recover = _follower_recovery_state->is_active();
     }
 
+    // abort here to pretend we are in recovery if we would without abortion
+    if (unlikely(_append_entries_blocked)) {
+        vlog(
+          _ctxlog.warn,
+          "ignoring append_entries request from {}: blocked",
+          r.source_node());
+        co_return reply;
+    }
```
Why are we blocking in this particular place and not earlier, i would imagine a failure injection is a general mechanism that prevents the follower from receiving append entries. Here we add a context to it as it is in a particullar place. I think we should reconisder the way this failure injection is done and try to approch it with more generic mechanism rather than adding a single flag

bashtanov replied:
The comment tells it briefly. We do it here because
1) we need feedback from the follower to mark it as lagging on the leader
2) at the same time we don't want the follower to change its log state based on this request

We might block later, but not earlier.

In reality such a situation (known lag) can only happen if some but not all append_entries requests come through. We can model it with injected delays or random failures, but this will drive the test into the land of scale tests I'm afraid, as without large amounts of data and consistent performance it'll be prone to flakiness.

mmaslankaprv commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -490,6 +490,86 @@
             ]
         },
+        {
+            "path": "/v1/debug/partitions/{namespace}/{topic}/{partition}/block_raft_op/{op}",
```
when looking at this it seems to be that it would be better to use honebadger, the mechanism is already there

bashtanov commented:
/dt

bashtanov commented:
@mmaslankaprv as discussed I removed the tests for now

bashtanov commented:
/dt

bashtanov commented:
/dt

dotnwat commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +278,341 @@ void health_monitor_backend::abort_current_refresh() {
+    operator==(const partition_risk&, const partition_risk&)
+      = default;
+
+    friend inline constexpr partition_risk
```
these do not need to be marked inline

dotnwat commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +278,341 @@ void health_monitor_backend::abort_current_refresh() {
+    friend std::ostream& operator<<(std::ostream&, const partition_risk&);
+};
+
+struct partition_risk::c {
```
why name this type `c` -- it's too short to have any obvious meaning right?

bashtanov replied:
`c` is for constants here. Originally I tried an enum class, but it had some limitations I could not overcome (something with bool cast maybe? I don't quite remember).

dotnwat replied:
might as well call it `constants` -- it's not like it saves much typing, but in return, you get a meaningful name.

bashtanov replied:
I wasn't after a meaningful name here, I would be happy with constants being in `partition_risk` directly, like it would be in an enum class.

My point was not to save seconds on typing but to maintain readability by preventing excessive wrapping.

dotnwat commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +278,341 @@ void health_monitor_backend::abort_current_refresh() {
+} // namespace
+
+ss::future<errc> health_monitor_backend::walk_local_and_remote_reports(
+  partition_leader_status_handler auto&& local_leader_handler,
+  partition_leader_status_handler auto&& remote_leader_handler,
+  partition_handler auto&& unclaimed_partition_handler) {
```
pass by value?

dotnwat commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +278,341 @@ void health_monitor_backend::abort_current_refresh() {
+        vlog(clusterlog.debug, "current node is not part of the cluster");
+        co_return errc::node_does_not_exists;
+    };
+    for (const auto& [nt, partitions] : local_report_it->second->topics) {
```
iteration is interleaving with scheduling points. is it safe?

bashtanov replied:
Thank you. It doesn't seem to be. I'll either replcae `_refresh_mutex` by an rwlock, or store reports behind an `lw_shared_ptr`.

dotnwat commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +278,341 @@ void health_monitor_backend::abort_current_refresh() {
+    }
+
+    // remote reports
+    for (const auto& [node_id, node_report] : _reports) {
+        if (node_id == _self) {
+            continue;
+        }
+        for (const auto& [nt, partitions] : node_report->topics) {
+            co_await ssx::async_for_each_counter(
```
iteration is interleaved with scheduling points. is it safe?

dotnwat commented on src/v/cluster/health_monitor_backend.cc:
```diff
@@ -271,6 +278,341 @@ void health_monitor_backend::abort_current_refresh() {
+health_monitor_backend::get_current_node_in_sync_replicas_share(
+  model::timeout_clock::time_point deadline) {
+    auto holder = _gate.hold();
+    if (
+      auto ec = co_await maybe_refresh_cluster_health(
+        force_refresh::no, deadline)) {
```
please use the `if (ec = ...; ec)` form. we should turn on warnings for the old form

bashtanov replied:
Is it an agreed style decision? What are its benefits?

dotnwat replied:
assignment-is-expression replaced by https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0305r0.html#motivation

bashtanov replied:
I'm very much a fan of the new style when the boolean expression is different from the initialized variable. But what's the point in repeating it after a semicolon if they are the same?

BTW there is an example in the proposal you linked:
`if (int s; int count = ReadBytesWithSignal(&s)) { publish(count); raise(s); }`.
Do you think it rather should be like this
`if (int s; int count = ReadBytesWithSignal(&s); count) { publish(count); raise(s); }`?

dotnwat commented on src/v/ssx/async_algorithm.h:
```diff
@@ -299,4 +300,76 @@ ss::future<> async_for_each_counter(
+    std::end(c);
+}
+ss::future<>
+async_for_each_counter(async_counter& counter, Container&& container, Fn f) {
+    if constexpr (std::is_lvalue_reference_v<decltype(container)>) {
+        return detail::async_for_each_fast<Traits>(
+          counter, std::begin(container), std::end(container), std::move(f));
+    } else {
+        return ss::do_with(
+          std::move(container),
+          [&counter, f = std::move(f)](Container& container) {
```
@bashtanov It seems like a footgun having different lifetime semantics depending on the value category. Why not pick one of the two and let the caller deal with it? For example it's very common for a caller to introduce the `do_with` if needed (or put some data on the stack and use co_await). Maybe i'm missing something here? cc @mmaslankaprv

bashtanov replied:
I think it's quite natural:
* if a caller provides an lvalue they do not expect us to move or copy from it, so it remains for them to hold it;
* if a caller provides an rvalue they expect us to take ownership of the data, including storing it for as long as necessary.
I even thought a `do_with` flavour with this logic would be quite useful. I would say it's more natural than silently making a copy.

> Why not pick one of the two and let the caller deal with it?

To reduce boilerplate for the convenience of the callers.

dotnwat replied:
I don't disagree with your analysis, but I think the thing that is being discounted here is consistency in behavior. AFAICT this is not consistent with most seastar/ssx APIs, right?

bashtanov replied:
Thanks, your point makes sense to me. Luckily all existing use of these two functions (there's one more above) will remain valid if we remove the `do_with`, as they all are co-awaited straight away. I'll open a thread in #cpplang tomorrow to gather more opinions.
